src/Simplex/Messaging/Agent.hs

```diff
@@ -141,13 +145,111 @@ resumeAgentClient c = atomically $ writeTVar (active c) True
 -- |
 type AgentErrorMonad m = (MonadUnliftIO m, MonadError AgentErrorType m)
 
+-- | Create SMP agent connection (NEW command) asynchronously, synchronous response is new connection id
+createConnectionAsync :: forall m c. (AgentErrorMonad m, ConnectionModeI c) => AgentClient -> Bool -> SConnectionMode c -> m ConnId
+createConnectionAsync c enableNtfs cMode = withAgentEnv c $ newConnAsync c enableNtfs cMode
+
+-- | Join SMP agent connection (JOIN command) asynchronously, synchronous response is new connection id
+joinConnectionAsync :: AgentErrorMonad m => AgentClient -> Bool -> ConnectionRequestUri c -> ConnInfo -> m ConnId
+joinConnectionAsync c enableNtfs = withAgentEnv c .: joinConnAsync c enableNtfs
+
+-- | Allow connection to continue after CONF notification (LET command), no synchronous response
+allowConnectionAsync :: AgentErrorMonad m => AgentClient -> ConnId -> ConfirmationId -> ConnInfo -> m ()
+allowConnectionAsync c = withAgentEnv c .:. allowConnectionAsync' c
+
+-- | Acknowledge message (ACK command) asynchronously, no synchronous response
+ackMessageAsync :: forall m. AgentErrorMonad m => AgentClient -> ConnId -> AgentMsgId -> m ()
+ackMessageAsync c = withAgentEnv c .: ackMessageAsync' c
+
+enqueueCommand :: forall m. AgentMonad m => AgentClient -> ConnId -> Maybe SMPServer -> ACommand 'Client -> m ()
+enqueueCommand c connId server aCommand = do
+  resumeCommandProcessing c server
+  commandId <- withStore c $ \db -> runExceptT . liftIO $ createCommand db connId server aCommand
+  queuePendingCommands c server [commandId]
+
+resumeCommandProcessing :: forall m. AgentMonad m => AgentClient -> Maybe SMPServer -> m ()
+resumeCommandProcessing c server =
+  unlessM (queueProcessing c server) $
+    async (runCommandProcessing c server)
+      >>= \a -> atomically (TM.insert server a $ asyncCmdProcesses c)
+
+resumeConnCommandsProcessing :: forall m. AgentMonad m => AgentClient -> ConnId -> m ()
+resumeConnCommandsProcessing c connId = do
+  unlessM connQueued $ do
+    cmdIdsSrvs <- withStore' c (`getPendingCommands` connId)
+    let srvCmdIds = groupCmdsBySrvs cmdIdsSrvs
```

Comment:
you could just read the list of servers, and then read commands for each server.

Reply:
not suggesting to change, some late thought...